### ISSUE TITLE
how to log TrainingObserver value

### ISSUE THREAD

**Issue by ccylance** [dmlc/xgboost#9648]
From the source code, I noticed there has TrainingObserver function to print Predictions, Gradients to print internal data structures. However, for python or R it seems there has no params to use the Observer to logging infos, so how could I logging these internal data structures like gain, best gain, weight, predict label values for each tree or even each node and each level during the training?

**Comment by trivialfis** [comment]
It's an internal debugging tool, you will have to change this option https://github.com/dmlc/xgboost/blob/0ecb4de9637ae1682334466d5a2cca69cb373ac8/CMakeLists.txt#L54 or modify the code to enable it. The output can be overwhelming though.

**Comment by ccylance** [comment]
So I have to use camke remake build a new c++ exe file right？But how could
I use it in python？And when I git clone the latest version code and build
from the source on mac，it do not generate xgboost execution file in build
directory

Jiaming Yuan ***@***.***>于2023年10月11日 周三09:00写道：

**Comment by trivialfis** [comment]
I don't think that debugging tool is what you want.

>  logging these internal data structures like gain, best gain, weight, predict label values for each tree or even each node and each level during the training

- best_gain: use `Booster.get_dump` with `with_stats` set to true.
- weight: same as above.
- predict label values for each tree: use model slicing. See the `Booster.__getitem__` method and Python document on read the doc site.
- even each node and each level during the training: you have to modify xgboost for these, see document site on how to build xgboost.

**Comment by trivialfis** [comment]
Feel free to reopen if there are further questions.
